juju/juju
# [2.8] Request a NIC per unique space for MAAS instances

Prior to this change, the constraints mapping code in the MAAS provider
would iterate the list of endpoint bindings and request an interface for
each individual endpoint even thought they might all be bound to the
same space.

As a result, when MAAS was asked to synthesize a KVM node with those
constraints it would generate multiple NICs labeled after each endpoint
thus triggering LP1902878.

Instead of having two separate code paths (v1 and v2) that intersected
(getBindings) and then diverged again, the fix introduced by this commit
extracts the space requirement calculation logic into a separate method
that is invoked by StartInstance and whose results are passed into
acquireNode/acquireNode2. This new method calculates the unique set of
endpoints from the application bindings (if provided) and appends any
positive space constraints. Then it calculates the negative space set
and checks for clashes with the positive set.

After this change, various bits of logic have been made redundant and
therefore removed. In addition, new tests have been added for the
introduced space requirement calculation logic and the existing
AcquireNode tests have been cleaned up as the validation logic is
applied before they are invoked.

## QA steps

Setup you virtual maas instance with a KVM instance and create 3 spaces: space1, space2, space3.
Assign each space to a different subnet and ensure that **none** of your already composed KVM pods
have a NIC in spaces 1 and 2 (the goal is to have MAAS synthesize a pod for us based on the constraints)

```sh
$ juju model-config logging-config='<root>=INFO;juju.provider.maas=TRACE'
$ juju deploy apache2 --bind space1 --constraints spaces=space2,^space3
```

Then check the provisioned KVM pod in MAAS web UI and verify that you have two NICs, one on space1 and one on space2.
If you check the model logs, you should be able to spot the constraints being passed to maas (IDs will probably be different in your case): "interfaces=space=1;space=2;not_subnets=space=3"

Try the same steps with juju from snap and you should see a NIC per endpoint name.

For consistency, we should try the same with a V1 MAAS.

## Bug reference
https://bugs.launchpad.net/juju/+bug/1902878
